packages/devtools_app/lib/src/flutter/table.dart

```diff
@@ -0,0 +1,395 @@
+import 'dart:math';
+
+import 'package:flutter/material.dart';
+
+import '../table_data.dart';
+import '../trees.dart';
+
+/// A table that shows [TreeNode]s.
+///
+/// This takes in a collection of [data], and a collection of [ColumnData].
+/// It takes at most one [TreeColumnData].
+///
+/// The [ColumnData] gives this table information about how to size its
+/// columns, how to present each row of `data`, and which rows to show.
+///
+/// To lay the contents out, this table's [_TreeTableState] creates a flattened
+/// list of the tree hierarchy. It then adds the nesting level of the
+/// deepest row in [data] that is currently visible with the prescribed
+/// width of each row in the table.
+class TreeTable<T extends TreeNode<T>> extends StatefulWidget {
+  const TreeTable({
+    Key key,
+    @required this.columns,
+    this.treeColumn,
+    @required this.data,
+    @required this.keyFactory,
+  }) : super(key: key);
+
+  /// The columns to show in this table.
+  final List<ColumnData<T>> columns;
+
+  /// The column of the table to treat as expandable.
+  final TreeColumnData<T> treeColumn;
+
+  /// The tree structure of rows to show in this table.
+  final T data;
+
+  /// Factory that creates keys for each row in this table.
+  ///
+  /// It uses [PageStorageKey] to restore the table's state when re-opening
+  /// its page.
+  final String Function(T data) keyFactory;
+
+  @override
+  _TreeTableState<T> createState() => _TreeTableState<T>();
+}
+
+class _TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
+    with TickerProviderStateMixin {
+  List<T> _flattenedList = [];
+
+  /// The width of each column in this table, including the
+  /// appropriate indentation for nested rows.
+  ///
+  /// This is computed in [_computeTableWidth].
+  List<double> columnWidths = [];
+  double get tableWidth => columnWidths.reduce((x, y) => x + y);
+
+  AnimationController _resizeController;
+  Animation<double> _resizeAnimation;
+  Tween<double> _widthTween;
+
+  /// The width to assume for columns that don't specify a width.
+  static const defaultColumnWidth = 500.0;
+
+  /// The maximum height to allow for rows in the table.
+  static const defaultRowHeight = 42.0;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.data.expandCascading();
+    _flattenedList = _flattenExpandedTree();
+    columnWidths = _computeTableWidth();
+    _widthTween = Tween<double>(begin: tableWidth, end: tableWidth);
```

Comment:
Why are begin and end the same value here?

Reply:
We start the animation off at the full width of the table, and update the tween when resizing the table to start it off at the current value of the tween and take it to the new tableWidth.

Look at the refreshTree() method for where that happens.

I've noticed that I don't get much different behavior between having and not having this animation, so I've decided to pull it out.